Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx

```diff
@@ -744,3 +746,63 @@ TEST(ShapedImageNeighborhoodRange, ConstructorSupportsRValueShapeOffsets)
   ASSERT_EQ(
     (RangeType{ *image, location, std::vector<OffsetType>{1} }).size(), 1);
 }
+
+
+// Tests that an arbitrary (possibly non-zero) index of the buffered region is supported.
+TEST(ShapedImageNeighborhoodRange, SupportsArbitraryBufferedRegionIndex)
+{
+  using ImageType = itk::Image<int>;
+
+  // Do zero-padding for neighborhood pixels outside the image:
+  using PolicyType = itk::Experimental::ConstantBoundaryImageNeighborhoodPixelAccessPolicy<ImageType>;
+  using RangeType = itk::Experimental::ShapedImageNeighborhoodRange<ImageType, PolicyType>;
+
+  // A minimal radius, and trivial shape offsets (n = 1), just for the test.
+  const itk::Size<ImageType::ImageDimension> radius = { {0} };
+  const auto offsets = itk::Experimental::GenerateRectangularImageNeighborhoodOffsets(radius);
+
+  const ImageType::IndexType zeroIndex{ {0} };
+  const ImageType::IndexType arbitraryIndex{ {10, -1} };
+  const ImageType::SizeType imageSize{ {2, 3} };
+
+  const ImageType::RegionType bufferedRegion{ arbitraryIndex, imageSize };
+
+  const auto image = ImageType::New();
+  image->SetRegions(bufferedRegion);
+  image->Allocate(true);
+
+  // Set a 'magic value' at the begin of the buffered region.
+  const ImageType::PixelType magicPixelValue = 42;
+  ImageType::PixelType* const bufferPointer = image->GetBufferPointer();
+  *bufferPointer = magicPixelValue;
+
+  RangeType range{ *image, zeroIndex, offsets };
+
+  // Expect zero, because zeroIndex is outside the buffered region.
+  EXPECT_EQ(range[0], 0);
+
+  range.SetLocation(arbitraryIndex);
+
+  // Expect the magic value as the first range value, because the range is now
+  // located at the arbitrary index, which corresponds to the begin of the
+  // image buffer.
+  EXPECT_EQ(range[0], magicPixelValue);
+
+  // Expect the magic value also when a range is constructed with
+  // the arbitrary index as its location.
+  EXPECT_EQ((RangeType{ *image, arbitraryIndex, offsets })[0], magicPixelValue);
+
+  // Check to see that this range yields the same pixel value as an
+  // itk::ConstNeighborhoodIterator with this arbitrary buffered region index.
+  using ConstNeighborhoodIteratorType =
+    itk::ConstNeighborhoodIterator<ImageType, itk::ConstantBoundaryCondition<ImageType>>;
+  ConstNeighborhoodIteratorType neighborhoodIterator(radius, image, image->GetRequestedRegion());
+
+  range.SetLocation(zeroIndex);
+  neighborhoodIterator.SetLocation(zeroIndex);
+  EXPECT_EQ(range[0], neighborhoodIterator.GetPixel(0));
```

Comment:
https://open.cdash.org/testDetails.php?test=712811110&build=5656366 reports a test failure: 

> Expected equality of these values:
>   range[0]
>     Which is: 0
>   neighborhoodIterator.GetPixel(0)
>     Which is: 175366032

The "old-school" `neighborhoodIterator.GetPixel(0)` appears to return an uninitialized value! What did I do wrong?

Reply:
Didn't read all, but have you initialized the image?
line 770:
```cpp
  const auto image = ImageType::New();
  image->SetRegions(bufferedRegion);
  image->Allocate(true);
  //FillBuffer(0)?
```

Reply:
Maybe the initialization of the whole image is not needed, but I guess zeroIndex should be initialized.
I guess `neighborhoodIterator.GetPixel(0)` pointer to the center of the neighborhood, ie, `zeroIndex`.

Reply:
@phcerdan Can you please have another look at the test that I wrote for this patch: https://github.com/InsightSoftwareConsortium/ITK/blob/f1d3a00f3a0fccd716d5c9ace78f5c3435b2bccf/Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx#L752 

The image type is `itk::Image<int>`, that should have all of its pixels zero-initialized by `image->Allocate(true)`, right?

With a buffered region of size 2 x 3, located at index [10, -1], `zeroIndex` should be outside of the buffered region, right?

So then the pixel value at  `zeroIndex` depends on the "boundary condition", right? In this case, I choose `itk::ConstantBoundaryCondition`, which should yield pixel value zero, by default, right?

I'm just asking because I could be wrong, of course! You know, I can make horrible bugs!

Reply:
Right! Good analysis. Not sure if that's a bug, or just we are breaking the preconditions of SetLocation.

The `itk::ConstNeighborhoodIterator::m_NeedToUseBoundaryCondition` gets set in the constructor inside the `ConstNeighborhoodIterator::SetRegion` method.
 With the test `ConstNeighborhoodIteratorType neighborhoodIterator(radius, image, image->GetRequestedRegion());`
the flag to use boundary conditions is correctly set to `false`. Try another region in the third argument (bigger, or not overlapping with the image), or maybe even setting it manually will do: `neighborhoodIterator.NeedToUseBoundaryConditionOn()`

Having said that, I am not sure if  `ConstNeighborhoodIterator::SetLocation(p)` has the precondition that the location p has to be inside the image. Try setting the flag manually, or change the region.

Reply:
@phcerdan Thanks for your analysis as well  `:-)`. Now I also think the test might be breaking a preconditions of `SetLocation`. (If so, does anyone know if this precondition is documented?) It could be that the function just was not designed to set the location outside of the image. Indeed, the test could be fixed by calling `neighborhoodIterator.NeedToUseBoundaryConditionOn()`, but that feels like cheating.  https://itk.org/ITKSoftwareGuide/html/Book1/ITKSoftwareGuide-Book1ch6.html#x51-1160006.4 says that that method _should be used with caution_... Later this week, I'll try to amend the test so that it would no longer violate this possible precondition.

Reply:
I think it is just an optimization flag, to not even check the boundaries, based on the regions of the constructor. No cheating setting it to `true`, I think it just enforces extra checkings like `IsInBounds`.